# Customer Churn Prediction

# Customer Churn Prediction using Machine Learning

## Objective
To predict whether a customer will leave the bank (Exited = 1) or continue using the bank's services (Exited = 0)
using machine learning algorithms.

In [3]:
# Importing libraries and dataset
import pandas as pd
df=pd.read_csv("Churn_Modelling.csv")
print(df.head())
print(df.shape)
print(df.info())
print(df.describe())

   RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         790

In [4]:
# Checking missing and duplicate values
print(df.isnull().sum())

print("Duplicate rows:", df.duplicated().sum())

RowNumber          0
CustomerId         0
Surname            0
CreditScore        0
Geography          0
Gender             0
Age                0
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          0
IsActiveMember     0
EstimatedSalary    0
Exited             0
dtype: int64
Duplicate rows: 0


In [5]:
# Checking Balance or imbalance dataset
print(df['Exited'].value_counts())

print(df['Exited'].value_counts(normalize=True) * 100)

Exited
0    7963
1    2037
Name: count, dtype: int64
Exited
0    79.63
1    20.37
Name: proportion, dtype: float64


In [6]:
# Checking whether balance of categorical columns
print(pd.crosstab(df['Geography'], df['Exited']))

print("\n")

print(pd.crosstab(df['Gender'], df['Exited']))

Exited        0    1
Geography           
France     4204  810
Germany    1695  814
Spain      2064  413


Exited     0     1
Gender            
Female  3404  1139
Male    4559   898


In [7]:
print(df.columns)

Index(['RowNumber', 'CustomerId', 'Surname', 'CreditScore', 'Geography',
       'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard',
       'IsActiveMember', 'EstimatedSalary', 'Exited'],
      dtype='str')


In [8]:
# Dropping irrelevant columns
df.drop("RowNumber", axis=1, inplace=True)
df.drop("CustomerId", axis=1, inplace=True)
df.drop("Surname", axis=1, inplace=True)

In [9]:
# Encoding categorical columns
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

df["Gender"] = le.fit_transform(df["Gender"])

In [10]:
# One-hot Encoding
df = pd.get_dummies(df, columns=["Geography"], drop_first=True)

In [14]:
df.head(10)
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CreditScore        10000 non-null  int64  
 1   Gender             10000 non-null  int64  
 2   Age                10000 non-null  int64  
 3   Tenure             10000 non-null  int64  
 4   Balance            10000 non-null  float64
 5   NumOfProducts      10000 non-null  int64  
 6   HasCrCard          10000 non-null  int64  
 7   IsActiveMember     10000 non-null  int64  
 8   EstimatedSalary    10000 non-null  float64
 9   Exited             10000 non-null  int64  
 10  Geography_Germany  10000 non-null  bool   
 11  Geography_Spain    10000 non-null  bool   
dtypes: bool(2), float64(2), int64(8)
memory usage: 800.9 KB


In [15]:
# Creating X & Y features
X = df.drop("Exited", axis=1)
y = df["Exited"]

In [16]:
print(X.shape)
print(y.shape)

(10000, 11)
(10000,)


In [17]:
# Train & test Split
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(8000, 11)
(2000, 11)
(8000,)
(2000,)


In [19]:
# Scaling
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

print(X_train_scaled.shape)
print(X_test_scaled.shape)

(8000, 11)
(2000, 11)


In [39]:
# Logistic Regression Model Training and Testing
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression(
    random_state=42)

lr.fit(X_train_scaled, y_train)

y_pred_lr = lr.predict(X_test_scaled)

In [40]:
# Confusion Matrix And Classification reports
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

print("Accuracy:", accuracy_score(y_test, y_pred_lr))

print(confusion_matrix(y_test, y_pred_lr))

print(classification_report(y_test, y_pred_lr))

Accuracy: 0.811
[[1543   64]
 [ 314   79]]
              precision    recall  f1-score   support

           0       0.83      0.96      0.89      1607
           1       0.55      0.20      0.29       393

    accuracy                           0.81      2000
   macro avg       0.69      0.58      0.59      2000
weighted avg       0.78      0.81      0.77      2000



In [43]:
# Decision Tree Classifier
from sklearn.tree import DecisionTreeClassifier

dt = DecisionTreeClassifier(
    random_state=42)

dt.fit(X_train, y_train)

y_pred_dt = dt.predict(X_test)

In [44]:
# Confusion Matrix And Classification reports
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

print("Accuracy:", accuracy_score(y_test, y_pred_dt))
print(confusion_matrix(y_test, y_pred_dt))
print(classification_report(y_test, y_pred_dt))

Accuracy: 0.7845
[[1371  236]
 [ 195  198]]
              precision    recall  f1-score   support

           0       0.88      0.85      0.86      1607
           1       0.46      0.50      0.48       393

    accuracy                           0.78      2000
   macro avg       0.67      0.68      0.67      2000
weighted avg       0.79      0.78      0.79      2000



In [49]:
# RandomForest Classifier
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(
    n_estimators=100,
    random_state=42
)

rf.fit(X_train, y_train)

y_pred_rf = rf.predict(X_test)

In [50]:
# Confusion Matrix And Clasfication reports
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

print("Accuracy:", accuracy_score(y_test, y_pred_rf))
print(confusion_matrix(y_test, y_pred_rf))
print(classification_report(y_test, y_pred_rf))

Accuracy: 0.866
[[1552   55]
 [ 213  180]]
              precision    recall  f1-score   support

           0       0.88      0.97      0.92      1607
           1       0.77      0.46      0.57       393

    accuracy                           0.87      2000
   macro avg       0.82      0.71      0.75      2000
weighted avg       0.86      0.87      0.85      2000



In [52]:
# Checking which feature contributes more in model performance
feature_importance = pd.DataFrame({
    'Feature': X.columns,
    'Importance': rf.feature_importances_
})

feature_importance = feature_importance.sort_values(
    by='Importance',
    ascending=False
)

print(feature_importance)

              Feature  Importance
2                 Age    0.239783
8     EstimatedSalary    0.146638
0         CreditScore    0.144214
4             Balance    0.138613
5       NumOfProducts    0.130255
3              Tenure    0.082427
7      IsActiveMember    0.041110
9   Geography_Germany    0.025711
1              Gender    0.019258
6           HasCrCard    0.018722
10    Geography_Spain    0.013268


In [54]:
# Comparision
comparison = pd.DataFrame({
    "Model": ["Logistic Regression", "Decision Tree", "Random Forest"],
    "Accuracy": [0.8110, 0.7845, 0.8675]
})

print(comparison)

                 Model  Accuracy
0  Logistic Regression    0.8110
1        Decision Tree    0.7845
2        Random Forest    0.8675


# Conclusion
In this project, Customer Churn Prediction was performed using the Churn_Modelling dataset containing 10,000 customer records.
The objective was to predict whether a customer would leave the bank (Exited = 1) or continue using its services (Exited = 0).

The dataset was preprocessed by removing irrelevant features such as RowNumber, CustomerId, and Surname.
Categorical variables were encoded using Label Encoding and One-Hot Encoding. The data was then split into training and testing sets,
and feature scaling was applied where required.

Three machine learning algorithms were implemented and evaluated:

| Model | Accuracy |
|---------|----------|
| Logistic Regression | 81.10% |
| Decision Tree | 78.45% |
| Random Forest | 86.75% |

Among all models, Random Forest Classifier achieved the best performance with an accuracy of 86.75%, along with better Precision,
Recall, and F1-Score compared to the other models. Feature importance analysis revealed that Age, EstimatedSalary, CreditScore, Balance,
and NumOfProducts were the most influential factors affecting customer churn.

The results demonstrate that machine learning techniques can effectively identify customers who are likely to leave the bank,
enabling organizations to take proactive customer retention measures and improve business performance.